librubyfmt/src/file_comments.rs

```diff
@@ -1,51 +1,54 @@
 use std::collections::BTreeMap;
-use std::io::{self, BufRead, Read};
 
 use crate::comment_block::CommentBlock;
+use crate::ruby::*;
 use crate::types::LineNumber;
 
-use regex::Regex;
-
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct FileComments {
     comment_blocks: BTreeMap<LineNumber, String>,
     contiguous_starting_indices: Vec<LineNumber>,
     lowest_key: LineNumber,
 }
 
 impl FileComments {
-    pub fn new() -> Self {
-        FileComments {
-            comment_blocks: BTreeMap::new(),
-            contiguous_starting_indices: vec![],
-            lowest_key: 0,
+    pub fn from_ruby_hash(h: VALUE) -> Self {
+        let mut fc = FileComments::default();
+        let keys;
+        let values;
+        unsafe {
+            keys = ruby_array_to_slice(rb_funcall(h, intern!("keys"), 0));
+            values = ruby_array_to_slice(rb_funcall(h, intern!("values"), 0));
+        }
+        if keys.len() != values.len() {
+            raise("expected keys and values to have same length, indicates error");
         }
+        for (ruby_lineno, ruby_comment) in keys.iter().zip(values) {
+            let lineno = unsafe { rubyfmt_rb_num2ll(*ruby_lineno) };
+            if lineno < 0 {
+                raise("line number negative");
+            }
+            let comment = unsafe { ruby_string_to_str(*ruby_comment) }
+                .trim()
+                .to_owned();
+            fc.push_comment(lineno as _, comment);
+        }
+        fc
     }
 
-    pub fn from_buf<R: Read>(r: io::BufReader<R>) -> io::Result<Self> {
-        lazy_static! {
-            static ref RE: Regex = Regex::new("^ *#").unwrap();
+    fn push_comment(&mut self, line_number: u64, l: String) {
+        if self.lowest_key == 0 {
+            self.lowest_key = line_number;
         }
-        let mut res = Self::new();
-        for (idx, line) in r.lines().enumerate() {
-            let l = line?;
-            if RE.is_match(&l) {
-                let line_number = (idx + 1) as LineNumber;
-                if res.lowest_key == 0 {
-                    res.lowest_key = line_number;
-                }
 
-                let last_line = res.contiguous_starting_indices.last();
+        let last_line = self.contiguous_starting_indices.last();
 
-                let should_push = line_number == 1
-                    || (last_line.is_some() && last_line.unwrap() == &(line_number - 1));
-                if should_push {
-                    res.contiguous_starting_indices.push(line_number);
-                }
-                res.comment_blocks.insert(line_number, l);
-            }
+        let should_push =
+            line_number == 1 || (last_line.is_some() && last_line.unwrap() == &(line_number - 1));
```

Comment:
```suggestion
        let should_push = last_line.copied().unwrap_or(0) == line_number - 1
```

Reply:
Or if my assumption that `last_line` is always `None` when `line_number == 1` is incorrect,
```suggestion
        let should_push =
            line_number == 1 || last_line.copied() == Some(line_number - 1);
```